File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
typedef long long ll;
ll solution(int n, vector<int> times) {
    ll answer = 0;
    ll min = 1;
    ll max = *max_element(times.begin(),times.end())*(ll)n;
    
    while(min<=max){
        ll mid = (min + max) / 2;
        ll people = 0;
        
        for(auto x: times){
            people += (mid/x);
        }
        if(people>=n){
            max = mid - 1;
            answer = mid;
        }
        else{
            min = mid + 1;
        }
    }
    
    
    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        long min = 1;
        long max = (long)Arrays.stream(times).max().getAsInt() * n;
        
        while(min <= max) {
            long mid = (min + max) / 2;
            long completed_people = 0;
            
            for(int t : times) {
                completed_people += (mid / t);
            }
            
            if(completed_people >= n) {
                max = mid - 1;
                answer = mid;
            }
            else {
                min = mid + 1;
            }    
        }
        return answer;
    }
}